### PR TITLE
RetroPlayer: OpenGL Plumbing

### PR DESCRIPTION
**Pull request by garbear**
## Description

This PR adds a lot of the plumbing that's been finished so far for OpenGL support in RetroPlayer.

It first extends RetroPlayer APIs to handle GL procs and framebuffer streams. Then it modifies the Game API to connect libretro cores up to this new functionality.

In the future, a new kind of renderbuffer will be implemented, an FBO renderbuffer. With this PR, it should be possible to implement the new renderbuffer type without further changes to the existing APIs or internal render manager implementation.

To complete GL support, we'll likely need modifications to rendering synchronization. That might end up being a pretty invasive change, but should be independent of the changes here. 

Requires a game.libretro PR: https://github.com/kodi-game/game.libretro/pull/129

## Motivation and context

OpenGL tracking issue: https://github.com/xbmc/xbmc/issues/17743

The code here is taken from the last time @lrusak and I worked on GL back in 2022: https://github.com/lrusak/xbmc/commits/retro-gl-v3/

The plumbing (up until the Game API bump in that branch) is relatively finished and complete, so I'll get it in master to reduce our technical debt and make it easier for future contributors to continue carrying the torch for GL in RetroPlayer.

I've prepared a branch that contains the additional FBO work gone into GL on top of this PR. It can serve as a starting place for future authors to pick up and advance the current work. Here's the branch: https://github.com/garbear/xbmc/commits/retro-gl-v3. Note that it includes a rebased #22211.

## How has this been tested?

Included in latest round of test builds: https://github.com/garbear/xbmc/releases/tag/retroplayer-21.1-20241213

Tested on my macbook M2. I tried some existing SW rendering cores, no regression were noted.

Next, I'll do further testing with GL cores. I suspect that many currently crash, which would be fixed by this PR. Before merging, I may also implement a dialog that notifies the user that GL support is WIP to handle the error more gracefully.

EDIT: I've done further testing on GL cores.

I verified that Mupen64-Next errs out when loading a game without crashing.

I tried Craft, and it also correctly aborts when HW rendering isn't available. However, it crashes when unloading - I'm preparing a patch for this. I've recorded the issue here: https://github.com/kodi-game/game.libretro.craft/issues/3

Screenshot 2024-12-13 at 9 27 05 PM

## What is the effect on users?

* Possibly fixes errors running unsupported OpenGL cores *(to be determined)*

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

(I'll check the "Bug fix" item if GL cores crash before this PR and are handled gracefully after)

### CODE REVIEW

**Comment by garbear** [comment]
I tested Parallel64. It doesn't fail without a framebuffer. It just happily runs without video. Savestates of size 16MB are successfully created.

Screenshot 2024-12-14 at 12 51 25 AM

**Comment by garbear** [comment]
I pushed a small update, which finally adds an error dialog when OpenGL games fail to play:

Screenshot 2024-12-14 at 3 50 03 PM

File: xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp [hunk 1]
```diff
@@ -265,6 +265,60 @@ void CRPRenderManager::AddFrame(const uint8_t* data,
...
+
+  for (auto renderBuffer : m_renderBuffers)
+    renderBuffer->Release();
+
+  m_renderBuffers = std::move(m_pendingBuffers);
+}
+
+uintptr_t CRPRenderManager::GetCurrentFramebuffer(unsigned int width, unsigned int height)
+{
+  std::vector<IRenderBuffer*> renderBuffers;
```
Comment (neo1973) [inline xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp:280]: Unused variable.

File: xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp [hunk 2]
```diff
@@ -265,6 +265,60 @@ void CRPRenderManager::AddFrame(const uint8_t* data,
...
+
+bool CRPRenderManager::Create(unsigned int width, unsigned int height)
+{
+  std::shared_ptr<CRPBaseRenderer> renderer = GetRendererForSettings(nullptr);
+  if (!renderer)
+    return false;
+
+  renderer->Configure(m_format);
+
+  std::vector<IRenderBuffer*> renderBuffers;
```
Comment (neo1973) [inline xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp:305]: Unused variable.

File: xbmc/cores/RetroPlayer/process/egl/RPProcessInfoEGL.h [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace KODI
+{
+namespace RETRO
+{
+class CRPProcessInfoEGL : public CRPProcessInfo
+{
+public:
+  CRPProcessInfoEGL(std::string platformName);
+  virtual ~CRPProcessInfoEGL() override = default;
```
Comment (neo1973) [inline xbmc/cores/RetroPlayer/process/egl/RPProcessInfoEGL.h:20]:
```suggestion
  ~CRPProcessInfoEGL() override = default;
```

**Review by neo1973 — CHANGES_REQUESTED** [review]
Not tested at all but looks good except some minor style stuff.

**Comment by garbear** [comment]
Thanks for the review @neo1973! I've pushed the fixups.

I now consider the PR to be code-complete. Before merging though, I want to add documentation outlining what is needed to finish the process. I plan to feed `o1` all convos, all code, all posts, and have it summarize down, and outline a plan of how to finish similar in form to a student proposal.

**Comment by garbear** [comment]
Here's the roadmap so far:

<hr>

# OpenGL Roadmap for RetroPlayer

OpenGL support in RetroPlayer will enable hardware rendering for cores that require OpenGL. This roadmap outlines the key tasks needed to complete its integration in Kodi.

The OpenGL tracking issue is: https://github.com/xbmc/xbmc/issues/17743

## Current Status

The foundational work for OpenGL support in RetroPlayer has been merged, including:

- **Plumbing APIs:** RetroPlayer APIs now support OpenGL procedures and framebuffer streams ([PR #26091](https://github.com/xbmc/xbmc/pull/26091))
- **Integration with Game API:** The Game API has been extended to connect libretro cores with OpenGL rendering functionality.
- **Core Compatibility:** Initial testing has verified that OpenGL cores, such as Mupen64Plus and Beetle PSX HW, can interface with RetroPlayer without crashing.

## Remaining Work

### VAOs and Rendering Pipeline

The use of Vertex Array Objects (VAOs) is a critical step in modernizing RetroPlayer's rendering pipeline. VAOs streamline the management of vertex attributes, making the OpenGL rendering process more efficient and less error-prone.

- **Current Progress:** The removal of global VAOs has been addressed in [PR #22211: Remove Global VAOs](https://github.com/xbmc/xbmc/pull/22211), which simplifies the rendering pipeline.
- **Next Step:** Review and merge [PR #22211](https://github.com/xbmc/xbmc/pull/22211) to finalize VAO updates.
- The PR has been rebased here: https://github.com/garbear/xbmc/commits/retro-gl-v3

Once the VAO work is completed, RetroPlayer will benefit from a cleaner and more maintainable rendering architecture.

### Frame Buffer Object (FBO) Renderbuffers

Mostly done. Original work [here](https://github.com/garbear/xbmc/commits/retro-gl-v3) and rebased work [here](https://github.com/garbear/xbmc/commits/retro-gl-v3).

<hr>

We also need to explain the challenges we were having with synchronization. I'll try to gather this info.

**Comment by garbear** [comment]
# OpenGL Roadmap for RetroPlayer

OpenGL support in RetroPlayer will enable hardware rendering for cores that require OpenGL. This roadmap outlines the key tasks needed to complete its integration in Kodi.

The OpenGL tracking issue is: https://github.com/xbmc/xbmc/issues/17743

## Current Status

The foundational work for OpenGL support in RetroPlayer has been merged, including:

- **Plumbing APIs:** RetroPlayer APIs now support OpenGL procedures and framebuffer streams ([PR #26091](https://github.com/xbmc/xbmc/pull/26091))
- **Integration with Game API:** The Game API has been extended to connect libretro cores with OpenGL rendering functionality.
- **Core Compatibility:** Initial testing has verified that OpenGL cores, such as Mupen64Plus and Beetle PSX HW, can interface with RetroPlayer without crashing.

## Remaining Work

### VAOs and Rendering Pipeline

The use of Vertex Array Objects (VAOs) is a critical step in modernizing RetroPlayer's rendering pipeline. VAOs streamline the management of vertex attributes, making the OpenGL rendering process more efficient and less error-prone.

- **Current Progress:** The removal of global VAOs has been addressed in [PR #22211: Remove Global VAOs](https://github.com/xbmc/xbmc/pull/22211), which simplifies the rendering pipeline.
- **Next Step:** Review and merge [PR #22211](https://github.com/xbmc/xbmc/pull/22211) to finalize VAO updates.
- The PR has been rebased here: https://github.com/garbear/xbmc/commits/retro-gl-v3

Once the VAO work is completed, RetroPlayer will benefit from a cleaner and more maintainable rendering architecture.

### Frame Buffer Object (FBO) Renderbuffers

Mostly done. Original work [here](https://github.com/garbear/xbmc/commits/retro-gl-v3) and rebased work [here](https://github.com/garbear/xbmc/commits/retro-gl-v3).

### Synchronization

The main synchronization challenges for OpenGL support in RetroPlayer stem from the separation of the game loop (which updates game state) and the rendering thread (which processes graphics).

- **GL Context Management:**
  OpenGL contexts are bound to threads, requiring careful management to ensure that rendering operations occur exclusively on  the rendering thread. Context sharing between threads needs to be avoided or strictly synchronized.

- **Dynamic Resolution Changes:**
  Dynamic resolution adjustments during gameplay add complexity. Both threads must coordinate to resize and reallocate frame buffers seamlessly, without causing crashes or visual glitches.

#### GL Context Management

Managing OpenGL contexts in RetroPlayer presents unique challenges due to its multithreaded architecture. Key considerations include:

- **Thread Binding:**
  OpenGL contexts are thread-specific. RetroPlayer must ensure that rendering occurs on a dedicated rendering thread without interfering with the game loop.

- **Context Resetting:**
  When switching contexts (e.g., during resolution changes), RetroPlayer must reinitialize OpenGL state while avoiding disruptions to ongoing operations.

- **Multiple Platforms:**
  Context management strategies must accommodate platform-specific OpenGL implementations, such as OpenGL ES for ARM-based devices or Metal wrappers for macOS.

To address these challenges, RetroPlayer can leverage context creation and management best practices, such as ensuring all rendering occurs on the same thread and using thread-safe proxy functions for OpenGL operations initiated by the game loop.

#### Dynamic Resolution Changes

Needed to accommodate gameplay scenarios where resolution changes are required mid-session, such as switching between fullscreen and windowed modes or adapting to performance constraints. This requires:

- **Real-Time Buffer Resizing:**
  Ensure frame buffers are resized and reallocated without disrupting gameplay or introducing visual artifacts.

- **Synchronization with Rendering Thread:**
  Both the game loop and the rendering thread must coordinate to apply resolution changes seamlessly.

---

## Conclusion

RetroPlayer's OpenGL support is well underway, with foundational work completed and key APIs in place. The remaining tasks include finalizing VAO updates, completing FBO renderbuffers, addressing synchronization challenges, and implementing dynamic resolution support. With these enhancements, RetroPlayer will achieve robust OpenGL integration, paving the way for future advancements like Vulkan support and broader hardware compatibility.

**Review by a1rwulf — APPROVED** [review]
I did not try to run it, but the code looks very clean now.
Thanks @garbear !

**Comment by sarbes** [comment]
Why is the removal of the global VAO a requirement? The renderer should get its own shared context.

**Review by neo1973 — APPROVED** [review]

**Pull request merged**